### PR TITLE
Add stopping rule StoppingLowestDoseHSRBeta

### LINKED ISSUE #111: New stopping rule that uses the beta distribution to stop for excessive tox at first dose
**Description**

Increment control is based on the number of observed DLTs and number of subjects at each dose level. The probability of toxicity is calculated using a Bin-Beta model with prior (a,b). If the probability exceeds the threshold for a given dose, that dose and all doses bove are excluded from further escalation.
This is a hard safety rule that limits further escalation based on the observed data per dose level.

**Additional Information**

Code will be provided in branch  111_StoppingLowestDoseHSRBeta

### PR DESCRIPTION
**Pull request by 0liver0815**
# Pull Request

Fixes #111

### CODE REVIEW

**Comment by 0liver0815** [comment]
The check R CMD Check failed as the warning
```
* checking Rd cross-references ... WARNING
Missing link or links in documentation object 'validate_stopping.Rd':
  ‘IncrementsHSRBeta’
```
pops up. I am not sure how to avoid the warning. The validation function for `validate_increments_hrs_beta` is the same as the one for `validate_stopping_lowest_dose_hsr_beta`. Therefore `validate_stopping_lowest_dose_hsr_beta` just calls the other. The first one is defined in another merge request and just repated here, but the cross-reference to the class is missing as the class is defined only in the other merge request.
Probably the reference can be removed, but this leads to different documentation for the same function (but in different merge requests).

**Comment by danielinteractive** [comment]
> The check R CMD Check failed as the warning
> 
> ```
> * checking Rd cross-references ... WARNING
> Missing link or links in documentation object 'validate_stopping.Rd':
>   ‘IncrementsHSRBeta’
> ```
> 
> pops up. I am not sure how to avoid the warning. The validation function for `validate_increments_hrs_beta` is the same as the one for `validate_stopping_lowest_dose_hsr_beta`. Therefore `validate_stopping_lowest_dose_hsr_beta` just calls the other. The first one is defined in another merge request and just repated here, but the cross-reference to the class is missing as the class is defined only in the other merge request. Probably the reference can be removed, but this leads to different documentation for the same function (but in different merge requests).

@0liver0815 in such a case I would suggest to either:
1) first merge the other PR that you depend on here, then continue here. (preferred usually)
2) in case the code in the 2 PRs is too inter-dependent -> better just make 1 PR.

**Comment by 0liver0815** [comment]
> > The check R CMD Check failed as the warning
> > ```
> > * checking Rd cross-references ... WARNING
> > Missing link or links in documentation object 'validate_stopping.Rd':
> >   ‘IncrementsHSRBeta’
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > pops up. I am not sure how to avoid the warning. The validation function for `validate_increments_hrs_beta` is the same as the one for `validate_stopping_lowest_dose_hsr_beta`. Therefore `validate_stopping_lowest_dose_hsr_beta` just calls the other. The first one is defined in another merge request and just repated here, but the cross-reference to the class is missing as the class is defined only in the other merge request. Probably the reference can be removed, but this leads to different documentation for the same function (but in different merge requests).
> 
> @0liver0815 in such a case I would suggest to either:
> 
> 1. first merge the other PR that you depend on here, then continue here. (preferred usually)
> 2. in case the code in the 2 PRs is too inter-dependent -> better just make 1 PR.

Thanks for your suggestion. Much appreciated, as this may also happen for future PR that are probably more inter-dependent. Here we can just merge #112 and then proceed with this one.

File: R/Rules-class.R [hunk 1]
```diff
@@ -686,7 +686,7 @@ IncrementMin <- function(IncrementsList)
 #' @seealso [`StoppingList`], [`StoppingCohortsNearDose`], [`StoppingPatientsNearDose`],
 #'   [`StoppingMinCohorts`], [`StoppingMinPatients`], [`StoppingTargetProb`],
 #'   [`StoppingMTDdistribution`], [`StoppingTargetBiomarker`], [`StoppingHighestDose`]
-#'   [`StoppingMTDCV`].
+#'   [`StoppingMTDCV`] [`StoppingLowestDoseHSRBeta`].
```
Comment (danielinteractive) [inline R/Rules-class.R:689]:
```suggestion
#'   [`StoppingMTDCV`], [`StoppingLowestDoseHSRBeta`].
```

File: R/Rules-class.R [hunk 2]
```diff
@@ -1037,6 +1037,78 @@ StoppingMTDCV <- function(target = 0.3,
...
+#' (i.e. above threshold probability) based on the observed data at the
+#' lowest dose level and a Beta(a,b) prior distribution.
+#' The default prior is Beta(1,1).
+#' In case that placebo is used, the rule is evaluated at the second dose of the
+#' dose grid, i.e. at the lowest non-placebo dose.
+#' Note: this stopping rule is independent from the underlying model.
+#'
+#' @slot target (`number`)\cr the target toxicity.
+#' @slot prob (`number`)\cr the threshold probability for the lowest
+#'  dose being toxic.
```
Comment (danielinteractive) [inline R/Rules-class.R:1060]:
```suggestion
#' @slot target (`proportion`)\cr the target toxicity.
#' @slot prob (`proportion`)\cr the threshold probability for the lowest
#'  dose being toxic.
```
even though the R class is numeric we relate to this as proportion here for the user and in the validation function

File: R/Rules-class.R [hunk 3]
```diff
@@ -1037,6 +1037,78 @@ StoppingMTDCV <- function(target = 0.3,
...
+  validity = validate_stopping_lowest_dose_hsr_beta
+)
+
+
+# StoppingLowestDoseHSRBeta-constructor ----
+
+#' @rdname StoppingLowestDoseHSRBeta-class
+#'
+#' @param target (`number`)\cr see slot definition.
+#' @param prob (`number`)\cr see slot definition.
```
Comment (danielinteractive) [inline R/Rules-class.R:1093]: same here

File: R/Rules-methods.R [hunk 4]
```diff
@@ -1587,6 +1587,65 @@ setMethod(
...
+  ),
+  definition = function(stopping, dose, samples, model, data, ...) {
+    # Actual number of patients at first active dose.
+    n <- sum(data@x == data@doseGrid[data@placebo + 1])
+
+    # Determine toxicity probability of the first active dose.
+    tox_prob_first_dose <-
+      if (n > 0) {
+        x <- sum(data@y[which(data@x == data@doseGrid[data@placebo + 1])])
+        1 - pbeta(stopping@target, x + stopping@a, n - x + stopping@b)
```
Comment (danielinteractive) [inline R/Rules-methods.R:1619]:
```suggestion
        pbeta(stopping@target, x + stopping@a, n - x + stopping@b, lower.tail = FALSE)
```

File: R/Rules-methods.R [hunk 5]
```diff
@@ -1587,6 +1587,65 @@ setMethod(
...
+      } else {
+        0
+      }
+
+    do_stop <- tox_prob_first_dose > stopping@prob
+
+    # generate message
+    msg <- if (n == 0) {
+      paste("Lowest active dose not tested, stopping rule not applied.")
```
Comment (danielinteractive) [inline R/Rules-methods.R:1628]:
```suggestion
      "Lowest active dose not tested, stopping rule not applied."
```

File: examples/Rules-class-StoppingLowestDoseHSRBeta.R [hunk 6]
```diff
@@ -0,0 +1,19 @@
+
+#   Stopping the study if the first dose is toxic with more than 90%
+#   probability based on a Beta posterior distribution with Beta(1,1) prior.
```
Comment (danielinteractive) [inline examples/Rules-class-StoppingLowestDoseHSRBeta.R:3]:
```suggestion
# Stopping the study if the first dose is toxic with more than 90%
# probability based on a Beta posterior distribution with Beta(1,1) prior.
```

File: examples/Rules-class-StoppingLowestDoseHSRBeta.R [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
```
Comment (danielinteractive) [inline examples/Rules-class-StoppingLowestDoseHSRBeta.R:4]:
```suggestion
```

File: examples/Rules-class-StoppingLowestDoseHSRBeta.R [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+my_stopping <- StoppingLowestDoseHSRBeta(
+  target = 0.3,
+  prob = 0.9
+)
+
+
+#   Stopping the study if the first dose is toxic with more than 90%
+#   probability based on a Beta posterior distribution with Beta(0.5,0.5) prior.
+
```
Comment (danielinteractive) [inline examples/Rules-class-StoppingLowestDoseHSRBeta.R:13]:
```suggestion
# Stopping the study if the first dose is toxic with more than 90%
# probability based on a Beta posterior distribution with Beta(0.5,0.5) prior.
```

File: tests/testthat/test-Rules-methods.R [hunk 9]
```diff
@@ -173,3 +173,132 @@ test_that("StoppingMTDCV works correctly if CV is above threshold", {
...
+    data = my_data
+  )
+  expected <- structure(
+    FALSE,
+    message = "Lowest active dose not tested, stopping rule not applied."
+  )
+  expect_identical(result, expected) # First active dose not applied.
+})
+
+my_data <- h_get_data_no_plcb()
```
Comment (danielinteractive) [inline tests/testthat/test-Rules-methods.R:247]: better move this in the tests. that makes it easier to run one test separately later

File: tests/testthat/test-Rules-methods.R [hunk 10]
```diff
@@ -173,3 +173,132 @@ test_that("StoppingMTDCV works correctly if CV is above threshold", {
...
+
+my_data <- h_get_data()
+
+# Model and samples are not needed to evaluate the rule, but needed for stopTrial
+# and therefore only generated once.
+my_model <- LogisticKadane(0.3, xmin = 0.001, xmax = 100)
+my_options <- McmcOptions(
+  burnin = 1, step = 1, samples = 1, rng_kind = "Mersenne-Twister", rng_seed = 94
+)
+my_samples <- mcmc(my_data, my_model, my_options)
```
Comment (danielinteractive) [inline tests/testthat/test-Rules-methods.R:188]: hm yeah I did not mention this above it seems, but in general this makes running individual tests harder. I would recommend
1) either use only one set of constant data sets defined on top of file. and then change/adapt inside each test
or
2) put this in a little helper function and then call this function within each of the tests below to recreate these objects, (e.g. return them in a list)

**Review by danielinteractive — COMMENTED** [review]
thx @0liver0815 ! please see last round of comments inside. looks nice!!

File: R/Rules-class.R (hunk 3 above)
Comment (0liver0815) [reply R/Rules-class.R:1093]: done

**Review by danielinteractive — APPROVED** [review]
thx @0liver0815 great work!

**Pull request merged**